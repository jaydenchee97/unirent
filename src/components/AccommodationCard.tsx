import { useNavigation } from "@react-navigation/native";
import { API, Auth, graphqlOperation } from "aws-amplify";
import dayjs from "dayjs";
import relativeTime from "dayjs/plugin/relativeTime";
import { useEffect, useState } from "react";
import { View, StyleSheet } from "react-native";
import {
  Card,
  Button,
  Text,
  Divider,
  Avatar,
  IconButton,
} from "react-native-paper";

import {
  createChatRoom,
  createUserChatRoom,
  updateUser,
} from "../graphql/mutations";
import IAccommodation from "../model/IAccommodation";
import { getCommonChatRoomWithUser } from "../services/ChatRoomService";
import {
  addSavedAccommodation,
  deleteSavedAccommodationById,
} from "../services/SavedAccommodationService";

dayjs.extend(relativeTime);

const AccommodationCard = (props: IAccommodation) => {
  const navigation = useNavigation();
  const [saved, setSaved] = useState(false);
  const [
    savedAccommodationAccommodationID,
    setSavedAccommodationAccommodationID,
  ] = useState("");

  const onContact = async () => {
    // check if have chatroom with user
    const existingChatRoom = await getCommonChatRoomWithUser(props.userId);
    if (existingChatRoom) {
      console.log("Chatroom already exists");
      navigation.navigate("Chat", { id: existingChatRoom.chatRoom.id });
      return;
    }

    console.log("Creating new chatroom");
    const newChatRoomData = await API.graphql(
      graphqlOperation(createChatRoom, { input: {} }),
    );
    if (!newChatRoomData.data?.createChatRoom) {
      console.log("Error creating new chat room");
    }
    const newChatRoom = newChatRoomData.data?.createChatRoom;

    // add clicked user to chatroom
    await API.graphql(
      graphqlOperation(createUserChatRoom, {
        input: { chatRoomId: newChatRoom.id, userId: props.userId },
      }),
    );

    // add auth user to chatroom
    const authUser = await Auth.currentAuthenticatedUser();
    await API.graphql(
      graphqlOperation(createUserChatRoom, {
        input: { chatRoomId: newChatRoom.id, userId: authUser.attributes.sub },
      }),
    );

    navigation.navigate("Chat", { id: newChatRoom.id });
  };

  const saveToggle = async () => {
    // toggle Saved function
    if (saved) {
      const deleted = await deleteSavedAccommodationById(
        savedAccommodationAccommodationID,
      );
      setSavedAccommodationAccommodationID("");
      if (deleted !== undefined) {
        setSavedAccommodationAccommodationID("");
      } else {
        console.log("delete save failed");
      }
    } else {
      const created = await addSavedAccommodation(
        props.savedAccommodationId,
        props.id,
      );
      if (created !== undefined) {
        setSavedAccommodationAccommodationID(
          created.data.createSavedAccommodationAccommodation.id,
        );
      } else {
        console.log("save failed");
      }
    }
    setSaved(!saved);
    if (props.onRerender !== undefined) {
      props.onRerender();
    }

    // setSavedAccommodationId(props.savedAccommodationId);
  };

  useEffect(() => {
    setSaved(!!props.isSaved);
    setSavedAccommodationAccommodationID(props.isSaved ? props.isSaved.id : "");
  }, [props]);

  return (
    <Card
      style={{
        flex: 1,
        backgroundColor: "#F4EDF9",
        borderColor: "#10161F",
        borderWidth: 0.3,
        marginVertical: 15,
      }}
      onPress={() =>
        navigation.navigate("Accommodation Detail", { id: props.id })
      }
      key={props.id}
    >
      <Card.Cover source={{ uri: props.images[0] }} />

      <View style={{ flex: 1, flexDirection: "row" }}>
        <Card.Title
          title={props.title}
          titleStyle={{ color: "gray" }}
          subtitle={props.address?.aptName}
          subtitleVariant="labelMedium"
          subtitleStyle={{ color: "gray" }}
          style={{ flex: 1 }}
        />
        <IconButton
          icon={saved ? "heart" : "heart-outline"}
          size={25}
          style={{
            marginVertical: 20,
            marginHorizontal: 20,
          }}
          onPress={async () => {
            await saveToggle();

            // setSaved(!saved);
          }}
        />
      </View>
      <Divider />
      <Card.Content style={{ marginVertical: 10 }}>
        <Text style={{ color: "gray" }}>
          S$ {props.price} / month â€¢ Available from {props.availableDate}
        </Text>
      </Card.Content>
      <Divider />
      <Card.Content style={{ marginVertical: 10 }}>
        <View style={{ flexDirection: "row", marginBottom: 10 }}>
          <Avatar.Text size={50} label="U" />
          <View
            style={{
              justifyContent: "center",
              marginHorizontal: 10,
              flexDirection: "column",
            }}
          >
            <Text style={{ fontSize: 12, fontWeight: "bold", color: "gray" }}>
              Listed By {props.User.name}
            </Text>
            <Text style={{ fontSize: 12, color: "gray" }}>
              {dayjs(props.createdAt).fromNow(true)} ago
            </Text>
          </View>
        </View>
        <Button mode="outlined" onPress={() => onContact()}>
          Contact
        </Button>
      </Card.Content>
    </Card>
  );
};

const styles = StyleSheet.create({
  blackFont: {
    color: "black",
  },
});

export default AccommodationCard;
