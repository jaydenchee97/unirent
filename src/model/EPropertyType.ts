enum EPropertyType {
  Condo = "Condo",
  Landed = "Landed",
  HDB = "HDB",
}

export default EPropertyType;
