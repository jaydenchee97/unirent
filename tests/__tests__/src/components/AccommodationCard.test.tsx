import renderer from "react-test-renderer";

import AccommodationCard from "../../../../src/components/AccommodationCard";
import EPropertyType from "../../../../src/model/EPropertyType";

describe("AccommodationCard", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <AccommodationCard
          id=""
          title=""
          propertyType={EPropertyType.Condo}
          price={0}
          description=""
          rented={false}
          address={undefined}
          availableDate=""
          images={[]}
          userId=""
          createdAt=""
          User={{ name: "test" }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
