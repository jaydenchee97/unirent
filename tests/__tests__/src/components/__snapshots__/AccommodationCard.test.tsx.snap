// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccommodationCard renders correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "#F4EDF9",
      "borderRadius": 12,
      "flex": 1,
      "marginVertical": 15,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="card-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#F4EDF9",
        "borderColor": "#10161F",
        "borderRadius": 12,
        "borderWidth": 0.3,
        "elevation": 1,
        "flex": undefined,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="card-container"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "flexShrink": 1,
          },
          undefined,
        ]
      }
      testID="card"
    >
      <View
        style={
          [
            {
              "backgroundColor": "#eeeeee",
              "height": 195,
              "overflow": "hidden",
            },
            {
              "borderRadius": 12,
            },
            undefined,
          ]
        }
      >
        <Image
          accessibilityIgnoresInvertColors={true}
          borderRadiusStyles={{}}
          siblings={
            [
              "Card.Cover",
              "View",
              "Card.Content",
              "Card.Content",
            ]
          }
          source={
            {
              "uri": undefined,
            }
          }
          style={
            [
              {
                "flex": 1,
                "height": undefined,
                "justifyContent": "flex-end",
                "padding": 16,
                "width": undefined,
              },
              {
                "borderRadius": 12,
              },
            ]
          }
        />
      </View>
      <View
        borderRadiusStyles={{}}
        index={1}
        siblings={
          [
            "Card.Cover",
            "View",
            "Card.Content",
            "Card.Content",
          ]
        }
        style={
          {
            "flex": 1,
            "flexDirection": "row",
          }
        }
        total={6}
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingLeft": 16,
              },
              {
                "minHeight": 50,
              },
              {
                "flex": 1,
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                },
              ]
            }
          />
          <View />
        </View>
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderRadius": 20.5,
              "height": 41,
              "margin": 6,
              "marginHorizontal": 20,
              "marginVertical": 20,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
              "width": 41,
            }
          }
          testID="icon-button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "transparent",
                "borderColor": "rgba(121, 116, 126, 1)",
                "borderRadius": 20.5,
                "borderWidth": 0,
                "elevation": 0,
                "flex": 1,
                "overflow": "hidden",
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="icon-button-container"
          >
            <View
              accessibilityComponentType="button"
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityTraits="button"
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              centered={true}
              collapsable={false}
              focusable={true}
              hitSlop={
                {
                  "bottom": 6,
                  "left": 6,
                  "right": 6,
                  "top": 6,
                }
              }
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  [
                    {
                      "alignItems": "center",
                      "flexGrow": 1,
                      "justifyContent": "center",
                    },
                    {
                      "borderRadius": 20.5,
                    },
                  ],
                ]
              }
              testID="icon-button"
            >
              <Text />
            </View>
          </View>
        </View>
      </View>
      <View
        borderRadiusStyles={{}}
        index={2}
        siblings={
          [
            "Card.Cover",
            "View",
            "Card.Content",
            "Card.Content",
          ]
        }
        style={
          [
            {
              "backgroundColor": "rgba(202, 196, 208, 1)",
              "height": 0.5,
            },
            undefined,
            false,
            false,
            undefined,
          ]
        }
        total={6}
      />
      <View
        borderRadiusStyles={{}}
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            undefined,
            {
              "marginVertical": 10,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "color": "gray",
              },
            ]
          }
        >
          S$ 
          0
           / month â€¢ Available from 
        </Text>
      </View>
      <View
        borderRadiusStyles={{}}
        index={4}
        siblings={
          [
            "Card.Cover",
            "View",
            "Card.Content",
            "Card.Content",
          ]
        }
        style={
          [
            {
              "backgroundColor": "rgba(202, 196, 208, 1)",
              "height": 0.5,
            },
            undefined,
            false,
            false,
            undefined,
          ]
        }
        total={6}
      />
      <View
        borderRadiusStyles={{}}
        style={
          [
            {
              "paddingHorizontal": 16,
            },
            {
              "paddingBottom": 16,
            },
            {
              "marginVertical": 10,
            },
          ]
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "marginBottom": 10,
            }
          }
        >
          <View
            style={
              [
                {
                  "backgroundColor": "rgba(103, 80, 164, 1)",
                  "borderRadius": 25,
                  "height": 50,
                  "width": 50,
                },
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                {},
              ]
            }
          >
            <Text
              numberOfLines={1}
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                  },
                  {
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "textAlign": "center",
                      "textAlignVertical": "center",
                    },
                    {
                      "color": "#ffffff",
                      "fontSize": 25,
                      "lineHeight": 25,
                    },
                    undefined,
                  ],
                ]
              }
            >
              U
            </Text>
          </View>
          <View
            style={
              {
                "flexDirection": "column",
                "justifyContent": "center",
                "marginHorizontal": 10,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                  },
                  {
                    "writingDirection": "ltr",
                  },
                  {
                    "color": "gray",
                    "fontSize": 12,
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              Listed By 
              test
            </Text>
            <Text
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                  },
                  {
                    "writingDirection": "ltr",
                  },
                  {
                    "color": "gray",
                    "fontSize": 12,
                  },
                ]
              }
            >
              a month
               ago
            </Text>
          </View>
        </View>
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderRadius": 20,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "transparent",
                "borderColor": "rgba(121, 116, 126, 1)",
                "borderRadius": 20,
                "borderStyle": "solid",
                "borderWidth": 1,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="button-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="button"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        [
                          {
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          false,
                          {
                            "marginHorizontal": 24,
                            "marginVertical": 10,
                          },
                          undefined,
                          false,
                          {
                            "color": "rgba(103, 80, 164, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          undefined,
                        ],
                      ],
                    ]
                  }
                  testID="button-text"
                >
                  Contact
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
